pkg/server/server.go

```diff
@@ -287,12 +284,149 @@ func (h *handler) authenticateEndpoint(w http.ResponseWriter, req *http.Request)
 	})
 }
 
-func renderTemplate(template string, identity *token.Identity) string {
+func (h *handler) doMapping(identity *token.Identity, arn string) (string, []string, error) {
+	if roleMapping, exists := h.lowercaseRoleMap[arn]; exists {
+		username, err := h.renderTemplate(roleMapping.Username, identity)
+		if err != nil {
+			return "", nil, fmt.Errorf("failed mapping username: %s", err.Error())
+		}
+		groups := []string{}
+		for _, groupPattern := range roleMapping.Groups {
+			group, err := h.renderTemplate(groupPattern, identity)
+			if err != nil {
+				return "", nil, fmt.Errorf("failed mapping group: %s", err.Error())
+			}
+			groups = append(groups, group)
+		}
+		return username, groups, nil
+	} else if userMapping, exists := h.lowercaseUserMap[arn]; exists {
+		return userMapping.Username, userMapping.Groups, nil
+	} else if _, exists := h.accountMap[identity.AccountID]; exists {
+		return identity.CanonicalARN, []string{}, nil
+	}
+	return "", nil, fmt.Errorf("ARN is not mapped: %s", arn)
+}
+
+func (h *handler) renderTemplate(template string, identity *token.Identity) (string, error) {
+
+	// Private DNS requires EC2 API call
+	if strings.Contains(template, "{{EC2PrivateDNSName}}") {
+		re := regexp.MustCompile("^i-(\\w{8}|\\w{17})$")
```

Comment:
I've seen `(^i-(\w{8}|\w{17})$)|(^mi-\w{17}$)` used elsewhere.  EC2 apparently also has [managed instances](https://docs.aws.amazon.com/systems-manager/latest/userguide/managed_instances.html) for hybrid environments with "mi-" id prefix.  Not sure if these should be allowed or not (are they covered by getPrivateDNSName?).

Reply:
@dbenhur Let's open a issue for this.  I think it requires more investigation to make sure there are no other side effects we aren't thinking about.

@nckturner Add the sign-off and I think this is good to go.

Reply:
Cool, will rebase and sign off.

Reply:
I think it'd also be fine to do minimal validation here and just trust that `ec2:DescribeInstances` is going to return an appropriate error if the instance ID is invalid.

We could also stick this regex in a global `var instanceIDPattern = regexp.MustCompile(...)` so we only compile once at init.